### ISSUE TITLE
classification accuracy?

### ISSUE THREAD

**Issue by FabianRoger** [CNuge/alfie#5]
Hi,

I was working with COI sequences from EMBL and BOLD and was trying to filter out miss-assigned sequences. I merged both datasets, clustered at 99% and the used alfie to predict the taxonomy of the centroid sequence. In total it's 574135 sequences. 

Here are my results  for the agreement (in %) between the Kingdom given by BOLD / EMBL (harmonized with the GBIF backbone) and alfie predictions. I was wondering what you make of this? For Fungi,  only 10 % of the sequences are classified? 

Note that the classified bold sequences vary in length, I will update the results with consistent fragment lengths. Interestingly, the embl data are barcoding fragments (bc of the way I extracted them with ecopcr) while BOLD seqs are as found on BOLD and vary greatly in length. 

Alfie is potentially super useful, so I don't mean to criticize in any way, just wondering what you though of the results below:

![image](https://user-images.githubusercontent.com/5137092/92935673-630fef00-f449-11ea-85ad-da98ffa902bf.png)

*edit* re-reading the preprint, I was wondering: How did you make sure the training data didn't contain miss-classifications? And if they did, would you expect that to have been a problem?

**Comment by CNuge** [comment]
Hi Fabian, 

Thanks for taking the time to run the alfie package on these data. Hopefully I can be of some help. If I'm interpreting the  above table correctly, ('Same_prct') is those where the classification predicted by alfie matches the reported taxonomy?
Two things are jumping out at me: 1. the Same_prct valyes for Fungi and Plantae are both very low, but only for the BOLD datasets. 2. The nseq value for fungi is higher than the number of COI fungi sequences on the BOLD database (~1500 last time I checked in April-ish).

Did you obtain the BOLD sequences using the BOLD API? If so, there is a quirk in that API, where even when you specify 'COI' as the barcode of interest it returns all barcode records of for specimens with COI sequences in the database, not just their COI sequences (i.e. if a fungi specimen was barcoded for COI and for ITS, it will give you both rows). The low performance for the fungi and plant bold categories (of which there are many non-COI barcode records on BOLD) suggests this may be the issue.

I would therefore encourage you to check the input dataframe for the presence of non-COI sequences, if you're passing non-COI barcodes to alfie the predictions will be poor/random, as the default neural network is trained only to classify COI sequences. 

If I've not diagnosed the issue properly then feel free to send me an email (possibly with the data) and I can help problem solve accordingly.

**Comment by FabianRoger** [comment]
Hi Cam. Thanks for taking the time to have a look at this!

Yes you probably spotted the main culprit here. I had downloaded the data 'manually' and missed to check what marker gene I got (BOLD is so tightly linked to COI  that I didn't think of it). After trimming my sequences I also lost a lot of fungal and plant sequences, especially from BOLD, so this points to the same error. 

Thanks!

re: training your NN, how did you make sure that misclassified sequences in the training data were not a problem? Or are NN robust to that?

**Comment by CNuge** [comment]
Hi Fabian, 
Glad to be of help with that! Sorry I missed the second question. Regarding the presence of misclassifications in the training data, there were a few different double checks to try to prevent that. First all sequences used in training passed BOLD's quality control tests (no stop codons when translated using the taxa specific table, blast check etc.). Additionally all the animal sequences were double checked using [coil](https://github.com/CNuge/coil), an R package I've produced for just that purpose. Unfortunately due to difference between kingdoms (presence of introns for some non-animals) this check could not be applied to all the sequences for the other kingdoms.

Lastly I did a PCA to determine if 1. the k-mer frequencies were effectively separating out the classes and 2. to search for evidence of misclassified sequences. As you can see in the following plot, the k-mer features separate the kingdoms pretty well along the first two axes of variation. Note you can see a low level of putative misclassifications on this plot too (green dot in the sea of red likely a fungi sequence called an animal, and one or two blue dots in the yellow cloud indicative of plant sequences labelled as bacteria). So there is likely an extremely small number of misclassified sequences in the data. I elected not to bias the data through micromanagement and manual removal of points, as the handful of putative labelling errors are unlikely strongly impact the classifier's performance given the tens of thousands of correctly labelled examples.


![kingdom_pca](https://user-images.githubusercontent.com/17327173/93095980-cd0add00-f671-11ea-9723-eeceeba41426.jpeg)
